### PR TITLE
Add BYOC docker file

### PR DESCRIPTION
**Pull request by dineshg13**
### What does this PR do?

This PR adds a Dockerfile that customers can use for Bringing their OTel components.  It wouldn't be used in CI. We are working on public documentation for it. 

It may not be the final place for this docker file. 

See https://docs.google.com/document/d/12vhVpx0slF-ltjXQUB9ZuePHqZ_9sK0QwQkyfmG2O80/edit#heading=h.bk3iqh57jna1

[Workflow](https://docs.google.com/document/d/1_pEm1Wqvt0CVYuD_rNvQ4e06LO-erdXInKCKR8wn4M8/edit)

### CODE REVIEW

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Use the Ubuntu Slim AMD64 base image
+FROM ubuntu:24.04 AS builder
+
+# Set environment variables
+ENV DEBIAN_FRONTEND=noninteractive
+ENV GO_VERSION=1.22.0
+ENV AGENT_VERSION=7.56.0-rc.7
```
Comment (truthbk) [inline Dockerfiles/agent-ot/Dockerfile.byoc:7]: I would not hardcode here this; instead I would default to `main` and allow overriding of the AGENT_VERSION with a [build argument](https://docs.docker.com/build/guide/build-args/).

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+COPY manifest.yaml /workspace/datadog-agent-${AGENT_VERSION}/comp/otelcol/collector-contrib/impl/manifest.yaml
+
+# Generate the files
+RUN . venv/bin/activate && invoke collector.generate
+
+# Build the OTel agent
+RUN . venv/bin/activate && invoke otel-agent.build
+
+# Use the final Datadog agent image
+FROM datadog/agent-dev:nightly-ot-beta-main-jmx
```
Comment (truthbk) [inline Dockerfiles/agent-ot/Dockerfile.byoc:77]: I would make the tag here `nightly-ot-beta-main-jmx` also a build argument.

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+# Create and activate virtual environment, then install requirements
+RUN python3 -m venv venv && \
+    . venv/bin/activate && \
+    pip install --upgrade pip && \
+    pip install -r tasks/requirements.txt
+
+# Copy the manifest file
+COPY manifest.yaml /workspace/datadog-agent-${AGENT_VERSION}/comp/otelcol/collector-contrib/impl/manifest.yaml
+
```
Comment (truthbk) [inline Dockerfiles/agent-ot/Dockerfile.byoc:69]: TODO: add an intermediate step to validate the manifest provided is aligned and makes sense.

(this comment is just for our own reference, no need to tackle in this PR).

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 1 above)
Comment (truthbk) [reply Dockerfiles/agent-ot/Dockerfile.byoc:7]: Same for `GO_VERSION` let's allow overriding with a build argument.

**Review by truthbk — COMMENTED** [review]
Looks good, just some small feedback.

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 1 above)
Comment (songy23) [reply Dockerfiles/agent-ot/Dockerfile.byoc:7]: +1

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+# Generate the files
+RUN . venv/bin/activate && invoke collector.generate
+
+# Build the OTel agent
+RUN . venv/bin/activate && invoke otel-agent.build
+
+# Use the final Datadog agent image
+FROM datadog/agent-dev:nightly-ot-beta-main-jmx
+
+ENV AGENT_VERSION=7.56.0-rc.7
```
Comment (songy23) [inline Dockerfiles/agent-ot/Dockerfile.byoc:79]: Here as well

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 2 above)
Comment (dineshg13) [reply Dockerfiles/agent-ot/Dockerfile.byoc]: done

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 1 above)
Comment (dineshg13) [reply Dockerfiles/agent-ot/Dockerfile.byoc]: done

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 3 above)
Comment (dineshg13) [reply Dockerfiles/agent-ot/Dockerfile.byoc:70]: created task https://datadoghq.atlassian.net/browse/OTEL-1986

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+RUN mkdir /go
+ENV GOPATH=/go
+
+# Verify installations
+RUN go version && \
+    python3 --version && \
+    curl --version
+
+# Download and unpack source code
+#RUN curl -L https://github.com/DataDog/datadog-agent/archive/refs/tags/${AGENT_VERSION}.tar.gz -o datadog-agent-${AGENT_VERSION}.tar.gz && \
```
Comment (songy23) [inline Dockerfiles/agent-ot/Dockerfile.byoc:51]: Why is this commented out? Is this still needed?

Comment (dineshg13) [reply Dockerfiles/agent-ot/Dockerfile.byoc:51]: Once we have an agent release with the otel binary, we should use code of the agent version. We currently use the main branch, which can have flaky or unreleased features/code.

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ENV PATH="/usr/local/go/bin:${PATH}"
+RUN mkdir /go
+ENV GOPATH=/go
+
+# Verify installations
+RUN go version && \
+    python3 --version && \
+    curl --version
+
+# Download and unpack source code
```
Comment (songy23) [inline Dockerfiles/agent-ot/Dockerfile.byoc:50]:
```suggestion
# TODO: use released agent version once we have an agent release with the otel binary
# Download and unpack source code
```

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 5 above)
Comment (songy23) [reply Dockerfiles/agent-ot/Dockerfile.byoc:51]: OK, I think it makes sense to add comment about it here

File: Dockerfiles/agent-ot/Dockerfile.byoc [hunk 7]
```diff
@@ -0,0 +1,84 @@
+ARG AGENT_VERSION=nightly-ot-beta-main-jmx
+# Use the Ubuntu Slim AMD64 base image
+FROM ubuntu:24.04 AS builder
+
+# Set environment variables
+ARG AGENT_VERSION
+ENV DEBIAN_FRONTEND=noninteractive
+ARG GO_VERSION=1.22.0
```
Comment (sblumenthal) [inline Dockerfiles/agent-ot/Dockerfile.byoc:8]: This is already out of date, the version according to the `.go-version` [file](https://github.com/DataDog/datadog-agent/blob/main/.go-version) is `1.22.5`

Which raises a question about the maintainability of this approach as it currently stands. Maybe we should move the go install to after the agent codebase git checkout, and get the go version from this file?

**Review by songy23 — APPROVED** [review]
LGTM pending the Go version change

File: Dockerfiles/agent-ot/Dockerfile.byoc (hunk 7 above)
Comment (dineshg13) [reply Dockerfiles/agent-ot/Dockerfile.byoc]: update

**Review by sblumenthal — APPROVED** [review]

**Pull request merged**